### ISSUE TITLE
ICE when compiling a *.d.ts file

### ISSUE THREAD

**Issue by jrop** [babel/babel#10502]
## Bug Report

**Current Behavior**
An exception is thrown when attempting to transform a *.d.ts file here: https://github.com/babel/babel/blob/master/packages/babel-plugin-transform-typescript/src/index.js#L142
```
TypeError: Cannot read property 'name' of null
  at PluginPass.Program (node_modules/@babel/plugin-transform-typescript/lib/index.js:167:57)
  at newFn (node_modules/@babel/traverse/lib/visitors.js:195:21)
  at NodePath._call (node_modules/@babel/traverse/lib/path/context.js:53:20)
  at NodePath.call (node_modules/@babel/traverse/lib/path/context.js:40:17)
  at NodePath.visit (node_modules/@babel/traverse/lib/path/context.js:88:12)
  at TraversalContext.visitQueue (node_modules/@babel/traverse/lib/context.js:120:16)
  at TraversalContext.visitSingle (node_modules/@babel/traverse/lib/context.js:92:19)
  at TraversalContext.visit (node_modules/@babel/traverse/lib/context.js:148:19)
  at Function.traverse.node (node_modules/@babel/traverse/lib/index.js:96:17)
  at traverse (node_modules/@babel/traverse/lib/index.js:78:12)
```

**Input Code**

```typescript
export default function<T>(x: T): T;
```

**Expected behavior/code**

Babel should not process *.d.ts files, or should give a better error.

**Babel Configuration (.babelrc, package.json, cli command)**

```js
{
  "presets": [["@babel/preset-env", { "targets": { "node": 10 } }], "@babel/preset-typescript"],
  "plugins": ["babel-plugin-macros"]
}
```

**Environment**
- Babel version(s): [e.g. v6.0.0, v7.0.0-beta.34]
- Node 10 / Yarn 1.17.3
- OS: MacOS 10.14.6
- Monorepo: yes: yarn workspace
- How you are using Babel: cli: `./node_modules/.bin/babel -d lib -x '.ts' src`

**Possible Solution**

When I add this to my config, it works:
```
{
  ...
  "ignore": ["src/dbg.macro.d.ts"]
}

```

**Additional context/Screenshots**
Add any other context about the problem here. If applicable, add screenshots to help explain.

**Comment by milesj** [comment]
`*.d.ts` files are declaration files, not source files. They are not meant to be parsed by Babel, so you should exclude them.

**Comment by jrop** [comment]
@milesj Yep, if you notice, that is listed as my solution.  What was hard to track down was Babel threw a very cryptic error, and that I had to hunt to find out what the actual issue was.  In this case, it was a very simple fix.  I did not post this as a request to process declaration files, but it seems feasible that people will inadvertently run into this as well: it is not uncommon when using TypeScript to get into situations that necessitate creating declaration files alongside src files.

**Comment by jrop** [comment]
To be a little more explicit, let me list out the steps that got me into this predicament:

**Comment by jrop** [comment]
To be a little more explicit, let me list out the steps that got me into this predicament:

1. I wanted to write a custom macro, and have it sit alongside my source.  In this case, I needed to write the macro in straight up JavaScript: no TypeScript: create a `src/my.macro.js`
2. I wanted to tell TypeScript how the macro would be "typed", since I would be using it in expressions: create a `src/my.macro.d.ts`
3. As soon as I created `src/my.macro.d.ts`, Babel gave the error listed above, with no hints as to what caused the problem.

Keep in mind that my development workflow in encountering this was not as straightforward as the simplified example above, but I had edited a large number of other files along with creating the `*.d.ts` file, so I had no idea what change triggered the error.

Thankfully, since node is easy to debug, I fired up the debugger and set a breakpoint on the offending line in the babel codebase.  Once suspended there, I could fish around for the file information to see where the error was originating from.  As soon as I saw that it was a declaration file, it was easy to fix.
